### ISSUE TITLE
`DataCollection->toArray` breaks grouping

### ISSUE THREAD

**Issue by rrelmy** [statamic/v2-hub#1527]
### Expected behaviour
``` php
Entry::whereCollection($collection)
    ->groupBy(function () {return 'test';})
    ->toArrray()`
```
should return an array with a `test` key in it.
``` php
$expected = [
    'test' => [/* items */],
];
```

Like the samples in the [laravel documentation](https://laravel.com/docs/5.1/collections#method-groupby)

### Actual behaviour
``` php
$actual = [
    0 => [/* items */],
];
```

The `DataCollection->toArray` function removes all the keys by using `array_values`

### Workaround
Use the default laravel collection by using `collect()` on your custom statamic collection.
``` php
$iWillGroupAsExpected = collect($statamicCollection);
```

### Server Details
**Statamic Version:** 2.6.1

**Comment by rrelmy** [comment]
Generally it always removes the keys. Other useful methods like `keyBy` are affected too.

**Comment by JesseLeite** [comment]
Yeah, it seems that `array_values()` should be removed there to better match Laravel's functionality...  But something else in core could be relying on that `array_values()` call, so it'd have to be done carefully.
